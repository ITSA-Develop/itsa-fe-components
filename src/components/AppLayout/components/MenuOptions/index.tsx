
import { TExtendedMenuItem, TMenuMode } from "@/types";
import { Menu, MenuProps } from "antd";
import React from "react";

export interface MenuOptionsProps {
	collapsed: boolean;
	currentPath: string;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	items?: TExtendedMenuItem[];
	mode?: TMenuMode;
	openKeys?: string[];
	onOpenKeysChange?: (openKeys: string[]) => void;
}

const MenuOptions: React.FC<MenuOptionsProps> = ({
  collapsed,
  currentPath = "",
  onClickOptionMenu,
  items = [],
  mode = "inline",
  openKeys = [],
  onOpenKeysChange,
}) => {

  const handleMenuClick: MenuProps['onClick'] = (info) => {
    // Find the clicked item in the menu data
    const findMenuItem = (menuItems: TExtendedMenuItem[], key: string): TExtendedMenuItem | null => {
      for (const item of menuItems) {
        if (item.key === key) {
          return item;
        }
        if ('children' in item && item.children) {
          const found = findMenuItem(item.children as TExtendedMenuItem[], key);
          if (found) return found;
        }
      }
      return null;
    };

    const clickedItem = findMenuItem(items, info.key);
    if (clickedItem && onClickOptionMenu) {
      onClickOptionMenu({ key: info.key, item: clickedItem });
    }
  };

  return (
    <div className="menu-options">
      <Menu
        mode={mode}
        inlineCollapsed={collapsed}
        items={items}
        className="overflow-auto scrollbar-none border-none min-w-full"
        defaultSelectedKeys={[currentPath]}
        openKeys={openKeys}
        onOpenChange={onOpenKeysChange}
        onClick={handleMenuClick}
      />
    </div>
  );
};

export default MenuOptions;
