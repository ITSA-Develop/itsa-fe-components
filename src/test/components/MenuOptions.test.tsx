import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import MenuOptions, { MenuOptionsProps } from '../../components/AppLayout/components/MenuOptions';

vi.mock('antd', async () => {
	const actual = await vi.importActual<any>('antd');

	const strip = (p: any) => {
		if (!p) return p;
		return p;
	};

	const Menu = (props: any) => (
		<div data-testid="menu" data-props={JSON.stringify(strip(props))}>
			<pre>{JSON.stringify(strip(props))}</pre>
			{props.children}
		</div>
	);

	return {
		...actual,
		Menu,
	};
});

const renderMenuOptions = (props: Partial<MenuOptionsProps> = {}) => {
	const defaultProps: MenuOptionsProps = {
		items: [],
		collapsed: false,
		currentPath: '/',
	};

	return render(<MenuOptions {...defaultProps} {...props} />);
};

const mockMenuItems = [
	{ key: '1', label: 'Dashboard' },
	{ key: '2', label: 'Users' },
	{
		key: 'sub1',
		label: 'Settings',
		children: [
			{ key: '3', label: 'Profile' },
			{ key: '4', label: 'Security' },
		],
	},
];

describe('MenuOptions', () => {
	it('renders correctly with default props', () => {
		const { container } = renderMenuOptions();

		expect(screen.getByTestId('menu')).toBeInTheDocument();
		expect(container.querySelector('.menu-options')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('passes correct props to Antd Menu component', () => {
		const { container } = renderMenuOptions({
			items: mockMenuItems,
			currentPath: '/dashboard',
			collapsed: false,
			mode: 'inline',
		});

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.items).toEqual(mockMenuItems);
		expect(menuProps.defaultSelectedKeys).toEqual(['/dashboard']);
		expect(menuProps.mode).toBe('inline');
		expect(menuProps.inlineCollapsed).toBe(false);
		expect(container).toMatchSnapshot();
	});

	it('sets inlineCollapsed to true when collapsed prop is true', () => {
		const { container } = renderMenuOptions({
			collapsed: true,
			items: mockMenuItems,
		});

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.inlineCollapsed).toBe(true);
		expect(container).toMatchSnapshot();
	});

	it('uses currentPath as defaultSelectedKeys', () => {
		const currentPath = '/users/profile';
		const { container } = renderMenuOptions({
			currentPath,
			items: mockMenuItems,
		});

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.defaultSelectedKeys).toEqual([currentPath]);
		expect(container).toMatchSnapshot();
	});

	it('renders with empty items array', () => {
		const { container } = renderMenuOptions({ items: [] });

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.items).toEqual([]);
		expect(container).toMatchSnapshot();
	});

	it('renders when collapsed state changes', () => {
		const { container, rerender } = renderMenuOptions({
			collapsed: false,
			items: mockMenuItems,
			currentPath: '/dashboard',
		});

		let menuElement = screen.getByTestId('menu');
		let menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');
		expect(menuProps.inlineCollapsed).toBe(false);

		rerender(<MenuOptions collapsed={true} items={mockMenuItems} currentPath="/dashboard" />);

		menuElement = screen.getByTestId('menu');
		menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');
		expect(menuProps.inlineCollapsed).toBe(true);
		expect(container).toMatchSnapshot();
	});

	it('handles complex menu items with nested structure', () => {
		const complexItems = [
			{
				key: 'main1',
				label: 'Main Section 1',
				children: [
					{ key: 'sub1-1', label: 'Subsection 1.1' },
					{
						key: 'sub1-2',
						label: 'Subsection 1.2',
						children: [
							{ key: 'deep1', label: 'Deep Item 1' },
							{ key: 'deep2', label: 'Deep Item 2' },
						],
					},
				],
			},
			{ key: 'main2', label: 'Main Section 2' },
		];

		const { container } = renderMenuOptions({ items: complexItems });

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.items).toEqual(complexItems);
		expect(container).toMatchSnapshot();
	});

	it('renders correctly when all props are provided', () => {
		const testProps = {
			items: mockMenuItems,
			collapsed: true,
			currentPath: '/test-path',
		};

		const { container } = renderMenuOptions(testProps);

		const menuElement = screen.getByTestId('menu');
		const menuProps = JSON.parse(menuElement.getAttribute('data-props') || '{}');

		expect(menuProps.items).toEqual(mockMenuItems);
		expect(menuProps.inlineCollapsed).toBe(true);
		expect(menuProps.defaultSelectedKeys).toEqual(['/test-path']);
		expect(menuProps.mode).toBe('inline');
		expect(container).toMatchSnapshot();
	});
});
