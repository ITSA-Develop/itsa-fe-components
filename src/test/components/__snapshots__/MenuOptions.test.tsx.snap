// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuOptions > handles complex menu items with nested structure 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":false,"items":[{"key":"main1","label":"Main Section 1","children":[{"key":"sub1-1","label":"Subsection 1.1"},{"key":"sub1-2","label":"Subsection 1.2","children":[{"key":"deep1","label":"Deep Item 1"},{"key":"deep2","label":"Deep Item 2"}]}]},{"key":"main2","label":"Main Section 2"}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":false,"items":[{"key":"main1","label":"Main Section 1","children":[{"key":"sub1-1","label":"Subsection 1.1"},{"key":"sub1-2","label":"Subsection 1.2","children":[{"key":"deep1","label":"Deep Item 1"},{"key":"deep2","label":"Deep Item 2"}]}]},{"key":"main2","label":"Main Section 2"}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > passes correct props to Antd Menu component 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":false,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/dashboard"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":false,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/dashboard"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > renders correctly when all props are provided 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/test-path"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/test-path"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > renders correctly with default props 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":false,"items":[],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":false,"items":[],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > renders when collapsed state changes 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/dashboard"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/dashboard"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > renders with empty items array 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":false,"items":[],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":false,"items":[],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > sets inlineCollapsed to true when collapsed prop is true 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":true,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;

exports[`MenuOptions > uses currentPath as defaultSelectedKeys 1`] = `
<div>
  <div
    class="menu-options"
  >
    <div
      data-props="{"mode":"inline","inlineCollapsed":false,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/users/profile"],"openKeys":[]}"
      data-testid="menu"
    >
      <pre>
        {"mode":"inline","inlineCollapsed":false,"items":[{"key":"1","label":"Dashboard"},{"key":"2","label":"Users"},{"key":"sub1","label":"Settings","children":[{"key":"3","label":"Profile"},{"key":"4","label":"Security"}]}],"className":"overflow-auto scrollbar-none border-none min-w-full","defaultSelectedKeys":["/users/profile"],"openKeys":[]}
      </pre>
    </div>
  </div>
</div>
`;
